### PR TITLE
feat: check GitHub check runs results before deployment

### PR DESCRIPTION
**Pull request by catcombo**
1. Replace `check suites` in a pre-deployment checks with a [GitHub check runs](https://docs.github.com/en/rest/reference/checks#list-check-runs-for-a-git-reference). `conclusion` value of all check runs should be `success` to make a deployment.
2. `ghd` will exclude check runs with the name `rebase` and with `skipped` conclusions by default. Use `--exclude-check-run-name` and/or `--exclude-check-run-conclusion` to override these values and filter out unintended check runs.

~Maybe instead of providing default values for `--exclude-check-run-name` and `--exclude-check-run-conclusion` we should add flags with default values to the [ghd script](https://github.com/moneymeets/ghd/blob/master/ghd#L24)? Because if there are default values, you cannot provide an empty list for these flags if necessary.~

### CODE REVIEW

File: commands/utils.py [hunk 1]
```diff
@@ -67,6 +67,8 @@ async def deploy(
     production: bool,
     description: str,
     check_constraints: bool,
+    exclude_check_run_name: list[str],
+    exclude_check_run_conclusion: list[str],
```
Comment (pchorus) [inline commands/utils.py:71]: I wonder if parameter names should be in plural form, like `exclude_check_run_names` and `exclude_check_run_conclusions`.

**Review by pchorus — COMMENTED** [review]
I added one minor question.

Furthermore, I like your proposal to avoid default values and instead add them to the script.
The important thing for me is that I can call `ghd gui` without providing any extra arguments.

File: commands/__init__.py [hunk 2]
```diff
@@ -184,6 +184,22 @@ class Columns(Enum):
     default=True,
     help="Check constraints before deployments, e.g. environment restrictions",
 )
+@click.option(
+    "-en",
```
Comment (marns93) [inline commands/__init__.py:188]: I would remove these shortcuts. In my opinion it is not clear enough and I would use the long name.

**Review by marns93 — CHANGES_REQUESTED** [review]
I like your proposal, too.
The default value can be an empty list and we can provide the flags in the script with our defaults.

File: commands/utils.py [hunk 3]
```diff
@@ -78,8 +80,24 @@ async def deploy(
                 ORDERED_ENVIRONMENTS,
             )
 
-        check_suites = await gh.get_check_suites(ref)
-        if not all(check["conclusion"] == "success" for check in check_suites if check["app"]["slug"] != "dependabot"):
+        # Collect check runs from all pages
+        check_runs, total_count = await gh.get_check_runs(ref, page=1)
+        for runs, _ in await asyncio.gather(
+            *(gh.get_check_runs(ref, page=page) for page in range(total_count // len(check_runs))),
```
Comment (mm-matthias) [inline commands/utils.py:86]: I think the current code is broken, e.g. for total_count = 50 you will do `for page in range(1)` or am I wrong?

What about this?
```suggestion
            total_pages = math.ceil(total_count / per_page)
            *(gh.get_check_runs(ref, page=page) for page in range(2, total_pages + 1)),
```
So sample cases:
- total_count=0 -> range(2, 2)
- total_count=50 -> range(2, 2)
- total_count = 100 -> range(2, 2)
- total_count = 101 -> range(2, 3)

File: github/__init__.py [hunk 4]
```diff
@@ -227,12 +227,9 @@ async def repositories(self) -> list[Repository]:
                 except KeyError:
                     return result
 
-    async def get_check_suites(self, ref: str) -> list:
-        # https://docs.github.com/en/rest/reference/checks#list-check-suites-for-a-git-reference
-        response = await self.get(f"/repos/{self.repo_path}/commits/{ref}/check-suites")
+    async def get_check_runs(self, ref: str, per_page: int = 100, page: int = 1) -> tuple[list, int]:
```
Comment (mm-matthias) [inline github/__init__.py:230]: I would probably make this 
```suggestion
    async def get_check_runs(self, ref: str) -> tuple[list, int]:
```
and move the paging code into this function. Then it is easy and clear to use imo.

**Review by mm-matthias — CHANGES_REQUESTED** [review]

File: commands/utils.py (hunk 3 above)
Comment (catcombo) [reply commands/utils.py]: > I think the current code is broken, e.g. for total_count = 50 you will do `for page in range(1)` or am I wrong?

You're right, I messed up with the old version of the code. Fixed.

File: github/__init__.py (hunk 4 above)
Comment (catcombo) [reply github/__init__.py:230]: I don't get it. You can't paginate without providing `per_page` and `page` with GitHub request.

File: commands/__init__.py (hunk 2 above)
Comment (catcombo) [reply commands/__init__.py:188]: I agree.

File: github/__init__.py (hunk 4 above)
Comment (mm-matthias) [reply github/__init__.py:230]: Yeah, it would do the paging internally in the function and return all check runs. Then you don't need the parameters.

File: commands/utils.py (hunk 1 above)
Comment (catcombo) [reply commands/utils.py:71]: I can't do that because the `click` library passes flag values as a keyword argument with the same name as the flag name. And I can't use plural form for the flags because `click` doesn't support [multi value options](https://click.palletsprojects.com/en/8.0.x/options/#multi-value-options) with variable number of values.

File: github/__init__.py (hunk 4 above)
Comment (catcombo) [reply github/__init__.py:230]: Ah, I got it. Good idea!

Comment (catcombo) [reply github/__init__.py]: Done.

File: commands/utils.py [hunk 5]
```diff
@@ -67,6 +67,8 @@ async def deploy(
     production: bool,
     description: str,
     check_constraints: bool,
+    exclude_check_run_name: list[str],
```
Comment (mm-matthias) [inline commands/utils.py:70]: I think the types here do not match. You use `tuple`s for the default values, but the type here is `list`. You might want to use e.g. `typing.Sequence` instead or stick to one of `tuple` or `list`.

**Review by mm-matthias — CHANGES_REQUESTED** [review]

File: github/__init__.py [hunk 6]
```diff
@@ -227,12 +228,23 @@ async def repositories(self) -> list[Repository]:
...
-            raise KeyError
+        async def _get_check_runs(per_page: int, page: int) -> tuple[list, int]:
+            # https://docs.github.com/en/rest/reference/checks#list-check-runs-for-a-git-reference
+            response = await self.get(
+                f"/repos/{self.repo_path}/commits/{ref}/check-runs?per_page={per_page}&page={page}",
+            )
+            GithubError.raise_from_message(response)
+            return response["check_runs"], response["total_count"]
+
+        check_runs, total_count = await _get_check_runs(per_page=100, page=1)
```
Comment (mm-matthias) [inline github/__init__.py:244]: Can you extract `100` into a variable? This makes it clear that all places where `100` appears actually have to match.

File: github/__init__.py [hunk 7]
```diff
@@ -227,12 +228,23 @@ async def repositories(self) -> list[Repository]:
...
+        for runs, _ in await asyncio.gather(
+            *(_get_check_runs(per_page=100, page=page + 2) for page in range(total_count // 100)),
```
Comment (mm-matthias) [inline github/__init__.py:246]: I think the new code is not 100% correct for the case `total_count=100`. It loop over `range(1)` which will then fetch `page=2`. But all 100 entries fit on page 1, so there is no need to fetch page=2 (maybe it is even an error, I don't know).

**Review by mm-matthias — CHANGES_REQUESTED** [review]

File: commands/utils.py (hunk 1 above)
Comment (catcombo) [reply commands/utils.py]: Shame on me. I don't know how to read the documentation. Just checked, it works. Code updated.

File: commands/utils.py (hunk 5 above)
Comment (catcombo) [reply commands/utils.py]: Done.

File: github/__init__.py (hunk 7 above)
Comment (catcombo) [reply github/__init__.py:246]: There is no error, GitHub return an empty list. But you're right. I need to fix the code according to that edge case.

File: github/__init__.py (hunk 6 above)
Comment (catcombo) [reply github/__init__.py]: I did it in the first version of code but then decided not to create a new variable in a small part of the code that will never be updated afterwards. Added.

**Comment by catcombo** [comment]
> Maybe instead of providing default values for `--exclude-check-run-name` and `--exclude-check-run-conclusion` we should add flags with default values to the [ghd script](https://github.com/moneymeets/ghd/blob/master/ghd#L24)? Because if there are default values, you cannot provide an empty list for these flags if necessary.

I just realized that I cannot do that because the `--exclude-check-run-name` and `--exclude-check-run-conclusion` flags are only relevant for `gui` and `deploy` sub-commands. I think it's hard to parse bash script parameters to extract the sub-command name and add the default flags only for the `gui` and `deploy` sub-commands.

**Review by mm-matthias — APPROVED** [review]

**Review by pchorus — APPROVED** [review]

**Review by marns93 — APPROVED** [review]
